ffi/derive/src/convert.rs

```diff
@@ -525,29 +762,130 @@ fn derive_into_ffi_for_fieldless_enum(enum_name: &Ident, repr: &[syn::NestedMeta
     }
 }
 
+fn derive_transparent(
+    name: &Ident,
+    generics: &syn::Generics,
+    item: &syn::DataStruct,
+) -> TokenStream2 {
+    let inner = item
+        .fields
+        .iter()
+        .next()
+        .map(|field| &field.ty)
+        .expect_or_abort("transparent struct must have at least one field");
+
+    let lifetime: syn::Lifetime = syn::parse_quote!('__iroha_ffi_itm);
+    let mut generics = generics.clone();
+    let (generic_params, ty_generics, where_clause) =
+        add_bounds_and_split_generics(&mut generics, &[], &[], &[]);
+
+    quote! {
+        // Note: if `#inner` is generic or doesn't have the same layout with `#name` this won't compile
+        impl <#generic_params> iroha_ffi::Transparent for #name #ty_generics #where_clause {
+            type Inner = #inner;
+
+            fn into_inner(outer: Self) -> Self::Inner {
+                unsafe { core::mem::transmute(outer) }
```

Comment:
why do you have to transmute? Couldn't you do something like `outer.0` or `outer.first_field_name`?

Reply:
I think i could do that quite easily in case of `IntoFfi`, but other way around would require additional assumptions like type has constructor function `new` or that every other field has default value.

Reply:
Not too keen on having `transmute` be generated by a macro. If it triggers UB, we need to be able to see it. 

As for the assumptions, I think that given this is done inside a macro, the exceptions that require transmute can be handled differently. At the very least it could be an argument to the derive macro, I'd prefer if there was a blanket `impl` and proper traits.

Reply:
> Not too keen on having transmute be generated by a macro. If it triggers UB, we need to be able to see it.

That's why `Transmute` trait must be made `unsafe` as I mentioned [here](https://github.com/hyperledger/iroha/pull/2644#discussion_r955754068)

Reply:
Made `Transparent` trait `unsafe`.